### ISSUE TITLE
Benchmark performance bottlenecks when opening chats

### ISSUE THREAD

**Issue by adhorodyski** [Expensify/App#50652]
Coming from the discussion thread under https://expensify.slack.com/archives/C05LX9D6E07/p1728314452317409

## Problem
Lack of solid performance benchmarks across the core app workflows, baselines for proposing performance improvements.

## Solution
- Focus on 1 scenario of **opening a chat**
- Benchmark it in detail to get the granular results (also for each phase) so we have solid numbers for whom specifically it's slow and why. There are also analytics we can use for this to have more datapoints.
- Pinpoint the areas for improvement we see
- Map how the proposal from initial thread/nitro modules/anything other can fit to improve it

**Comment by adhorodyski** [comment]
@mountiny could you please assign this one to me & @hannojg? Thanks!

**Comment by hannojg** [comment]
Thanks! 👋

**Comment by adhorodyski** [comment]
**Update**
Our bottom line is to get super solid on _what precisely is slow_ (have granular measurements) and _for whom specifically_.

**Who**
To be able to say how many people something affects (in different percentiles), we need to have something to segment by. We'll extend our performance analytics' metadata (in all envs, including e2e tests) to include the below fields. These will help us group measurements by the types of accounts.
- number of reports ✅
- number of reportActions 🔄
- number of personalDetails ✅
- number of policies 🔄
- number of transactions 🔄
- number of transactionViolations 🔄

**What**
To be able to pinpoint what exactly is slow, we'll implement granular metrics in code, across both App & Onyx. Here's a breakdown of all the phases we decided to focus on.
- **APPLYING_UPDATES** (coming over https, until we schedule notifying subscribers)
    - **DATABASE_WRITE** (actual save operations like multiset and multimerge)
- **NOTIFYING_SUBSCRIBERS** (multiple key(s)Changed triggered by scheduleNotifyCollectionSubscribers)
    - **DATABASE_READ**
    - **REACT_PROVIDER** - (useOnyx/withOnyx)
- **ONYX_DATA_POSTPROCESSING** (from receiving a new update (subscriber got called) until we have the final data)

**Comment by adhorodyski** [comment]
cc @mountiny we'd like to ask you for any feedback about this 1st batch (please invite others!). Is there an area we're missing? I want to make sure we focus on the right parts when profiling this, the exact tools (like e2e suites) will come next once we fully agree on the who&what.

**Comment by adhorodyski** [comment]
The extended analytics setup is merged now:
- to be confirmed it's available in Firebase for newly reported events
- to be merged into local-based profiling tools so we can use the same parameters in eg. e2e workflows cc @hannojg, I think this is the last thing we have to do when it comes to the 'segmentation' part

**Comment by adhorodyski** [comment]
- Next part is to implement the granular metrics as listed above

also cc @hannojg, do you think these should also be reported to Firebase OR we just focus on being able to profile this behind some flag?

**Comment by adhorodyski** [comment]
I'm back to this issue after some inactivity from my end, will post an update tomorrow :)

**Comment by adhorodyski** [comment]
For starters I'm investigating why updated metadata fields are not visible in Firebase. They're in code [here](https://github.com/callstack-internal/Expensify-App/blob/4735b5879a0dcc018ac8c68e94949999ad386daf/src/libs/Firebase/utils.ts#L9) for 2 weeks now.

**Comment by mountiny** [comment]
@adhorodyski are you getting staging data there too? latest staging is 9.0.55-9

**Comment by hannojg** [comment]
Sorry, this completely slipped through the cracks!

> do you think these [the granular metrics as listed above] should also be reported to Firebase OR we just focus on being able to profile this behind some flag?

I think behind some flags so we can test those locally. I think it might be a bit too much to track all those events and send them to analytics. Lets get those implemented! How can i help on those?

**Comment by adhorodyski** [comment]
I'm starting with the more high-level marks [here](https://github.com/Expensify/App/pull/51815), but for this very benchmark we need `Performance.markStart/End()` calls as described [above](https://github.com/Expensify/App/issues/50652#issuecomment-2416763503) so we can collect them to memory. Can you possibly pick even a few of them @hannojg?

**Comment by hannojg** [comment]
Yes, will check! This will also involve making a few changes in onyx to enable these kind of measurements!

**Comment by hannojg** [comment]
Working on a PR for onyx over here:

- https://github.com/Expensify/react-native-onyx/pull/598

expecting to wrap it up by tomorrow!

With these changes we should be able to record all steps of what onyx does using the performance API / react-native-performance. We can then use the same API in the app to create reports for different actions performed.

**Comment by adhorodyski** [comment]
Reviewing now!

**Comment by adhorodyski** [comment]
Approved the PR to Onyx. I'll start adding the other metrics' calls (app-specific) into the app next week.

**Comment by adhorodyski** [comment]
https://github.com/Expensify/App/pull/53278 here's another PR for the `applying_updates` stage.

**Comment by adhorodyski** [comment]
Here's how we're currently looking with 2 items left. Not sure how we should wrap these though yet.
- APPLYING_UPDATES 🛠️ PR by me
- DATABASE_WRITE 🛠️ PR by Hanno
- NOTIFYING_SUBSCRIBERS 🛠️ PR by Hanno
- DATABASE_READ 🛠️ PR by Hanno
- REACT_PROVIDER ⏭️
- ONYX_DATA_POSTPROCESSING ⏭️

I'm also thinking on how we can visualize this data. I can see this being exportable as any other trace (it's an array of measurements) so we can maybe build a Gant chart out of this.

**Comment by adhorodyski** [comment]
After some considerations around `ONYX_DATA_POSTPROCESSING` I think it'd be the most precise (and easy to follow) to actually wrap all specific computations separately. I'll look for places we know take a bit, wrap them with `markStart/markEnd` so we are certain which specific parts are slow.

**Comment by adhorodyski** [comment]
Hey, linking this small one: https://github.com/Expensify/App/pull/53652

**Comment by adhorodyski** [comment]
I'm now working on the last decorators within Onyx so we can have `withOnyx/useOnyx` wrapped as well.

**Comment by adhorodyski** [comment]
PR to Onyx on decorating `useOnyx` is there (wip): https://github.com/Expensify/react-native-onyx/pull/607

I'll test it & wrap the checklist on tomorrow.

**Comment by adhorodyski** [comment]
Just opened the PR for review here: https://github.com/Expensify/react-native-onyx/pull/607

**Comment by mountiny** [comment]
@adhorodyski Anything else to do here?

**Comment by mountiny** [comment]
Could you and @hannojg please try to write down all the next steps as a checklist? I think that will force us to see this in a grand scheme and keep you accountable too. Right now its not clear what is next

**Comment by muttmuure** [comment]
Yeah it would be great to know when we can see the first results from this

**Comment by adhorodyski** [comment]
Hey, sorry for missing an update there! Since the apps are putting the markers now in the right places, our next steps would be to:
- collect them from device (preferably to a file)
- build a way for understanding the results (big array of entries will be hard on its own)

1. This can be built very similar to the other debug exports we do, with a troubleshoot export action.
2. I think it'll even be enough to drop them into an LLM for analysis. These are not secrets but marked timestamps so we can take this shorter path I think.

**Comment by adhorodyski** [comment]
Opened a [draft PR](https://github.com/Expensify/App/pull/54412) and I will proceed with more testing so this can be reviewed from Monday. 
I just have to make sure it works as expected (if 'measures' need to be filtered in some cases), but it looks stable. I'll try analyzing this data at the same time.

**Comment by adhorodyski** [comment]
Tested this today, here's an update:

```
[{"detail": undefined, "duration": 15.607542037963867, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968390627.430583}, {"detail": undefined, "duration": 15.607542037963867, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968390627.430583}, {"detail": undefined, "duration": 15.607542037963867, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968390627.430583}, {"detail": undefined, "duration": 259.7711670398712, "entryType": "measure", "name": "sidebar_loaded", "startTime": 968390619.279541}, {"detail": undefined, "duration": 14.05049991607666, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968390969.205625}, {"detail": undefined, "duration": 0.24849998950958252, "entryType": "measure", "name": "apply_https_updates", "startTime": 968391216.804041}, {"detail": undefined, "duration": 0.3620830774307251, "entryType": "measure", "name": "apply_https_updates", "startTime": 968391539.429333}, {"detail": undefined, "duration": 0.11170804500579834, "entryType": "measure", "name": "apply_https_updates", "startTime": [968392154.4035](tel:9683921544035)}, {"detail": undefined, "duration": 18.054041028022766, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968393363.46525}, {"detail": undefined, "duration": 0.1734999418258667, "entryType": "measure", "name": "apply_https_updates", "startTime": 968393604.243}, {"detail": [Object], "duration": 69.57566821575165, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968393895.232208}, {"detail": [Object], "duration": 0.44399988651275635, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968394047.808}, {"detail": [Object], "duration": 1.517503023147583, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968394082.572625}, {"detail": undefined, "duration": 0.1409590244293213, "entryType": "measure", "name": "apply_https_updates", "startTime": 968393757.601416}, {"detail": undefined, "duration": 16.013083934783936, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968394016.605291}, {"detail": undefined, "duration": 1403.4333330392838, "entryType": "measure", "name": "open_report", "startTime": 968392672.306208}, {"detail": [Object], "duration": 15.28348159790039, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968394166.443416}, {"detail": [Object], "duration": 0.01708197593688965, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968394288.094958}, {"detail": undefined, "duration": 0.15479207038879395, "entryType": "measure", "name": "apply_https_updates", "startTime": 968394537.854291}, {"detail": [Object], "duration": 0.009166121482849121, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968395254.338}, {"detail": [Object], "duration": 21.432971477508545, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968396990.421416}, {"detail": [Object], "duration": 3.0215084552764893, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397078.794375}, {"detail": [Object], "duration": 0.4611673355102539, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397110.418458}, {"detail": undefined, "duration": 14.88450002670288, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968397057.165666}, {"detail": undefined, "duration": 14.88450002670288, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968397057.165666}, {"detail": undefined, "duration": 217.44841706752777, "entryType": "measure", "name": "send_message", "startTime": 968396858.680833}, {"detail": [Object], "duration": 0.01737511157989502, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397141.732083}, {"detail": [Object], "duration": 1.3695030212402344, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397200.349}, {"detail": [Object], "duration": 17.675146460533142, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397319.209125}, {"detail": [Object], "duration": 0.08849763870239258, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": [968397403.1635](tel:9683974031635)}, {"detail": undefined, "duration": 0.18762493133544922, "entryType": "measure", "name": "apply_https_updates", "startTime": 968397233.586125}, {"detail": undefined, "duration": 15.384583950042725, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968397382.722291}, {"detail": undefined, "duration": 15.384583950042725, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968397382.722291}, {"detail": [Object], "duration": 0.015044093132019043, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397442.718333}, {"detail": [Object], "duration": 0.01949894428253174, "entryType": "measure", "name": "<ReportActionsView> rendering", "startTime": 968397538.173208}, {"detail": undefined, "duration": 16.266417026519775, "entryType": "measure", "name": "get_ordered_report_ids", "startTime": 968398106.290208}, {"detail": undefined, "duration": 16.216791033744812, "entryType": "measure", "name": "apply_https_updates", "startTime": 968399287.737375}, {"detail": undefined, "duration": 0.16920804977416992, "entryType": "measure", "name": "apply_https_updates", "startTime": 968399688.226458}]
```

^ is a sample output we get from listing all the local `measures`. The format is really nice for LLMs to process, so we easily get such a summary:
```
Here are the top 5 longest operations from the performance logs, sorted by duration (in milliseconds):

open_report: 1,403.43 ms
sidebar_loaded: 259.77 ms
send_message: 217.45 ms
apply_https_updates: 16.22 ms (longest instance)
<ReportActionsView> rendering: 69.58 ms (longest instance)
```

My one problem I observed today though is that we only collect these behind an environment variable flag of `CAPTURE_METRICS`, so it won't have an effect in production and only produce a `[]` (my bad I tested this against a local override).

I thought about turning the markers on together with the `BaseProfilingToolMenu` available under `Use profiling` cc @hannojg.

**Comment by adhorodyski** [comment]
Working on integrating measures into https://github.com/Expensify/App/blob/8b7096fc4c4d522407a6f51b4fe0fd3543e5feb2/src/components/ProfilingToolMenu/BaseProfilingToolMenu.tsx#L53 the same way the Memoize is integrated so we can just turn this on/off.

**Comment by adhorodyski** [comment]
Still working on the linked draft PR, reworking the Performance module so it's available during runtime. I'm also adjusting it to fit our style guide.

**Comment by adhorodyski** [comment]
Updated [the PR](https://github.com/Expensify/App/pull/54412), testing this now.

**Comment by adhorodyski** [comment]
@mountiny the PR got merged but I think we're blocked on evaluating this further given the profiling feature itself is currently broken. Everything is in place though, we're not collecting local metrics to the AppInfo. Should we close this issue?

**Comment by muttmuure** [comment]
> the profiling feature itself is currently broken.

Broken in HybridApp? Or broken totally?

**Comment by mountiny** [comment]
The RN upgarde was merged so we can hopefully profile again very soon

**Comment by adhorodyski** [comment]
@muttmuure yeah that was totally I think due to this underlying bug in Hermes. I'll monitor when it's up. @mountiny the latest testflight includes upgrade, correct?

**Comment by mountiny** [comment]
Yes

**Comment by adhorodyski** [comment]
Ok so I've found a few bugs on different environments that I can't go over quickly:
- https://staging.new.expensify.com/ on web allows me to profile, but does not seem to produce the `AppInfo` file
- https://new.expensify.com/ on web does not allow me to profile, `Use profiling` does not toggle on with a console error of `[Violation] Document policy violation: js-profiling is not allowed in this document.`
- TestFlight build is similar to the web staging - allows me to profile, but does not produce the `AppInfo`

cc @hannojg are these known issues, am I missing something? Shouldn't AppInfo actually be produced on all these targets?

**Comment by adhorodyski** [comment]
I think there are no actions needed in the scope of this task, the feature works as expected though I have some doubts it's equally accessible across all the platforms (it haven't been touched here). Should we follow up on this one as per ^? cc @mountiny

**Comment by mountiny** [comment]
@adhorodyski if you mean this https://github.com/Expensify/App/issues/50652#issuecomment-2582796537 yes lets try to follow up

**Comment by laurenreidexpensify** [comment]
@rayane-djouah are you onboarded yet for payment?

**Comment by laurenreidexpensify** [comment]
@rayane-djouah have done, am bumping the team to get it resolved today 👍

**Comment by garrettmknight** [comment]
@laurenreidexpensify got the payment request in ND, just need the payment summary when you get a chance. 😄

**Comment by garrettmknight** [comment]
$250 approved for @rayane-d
